gridfs/grid_file.py

```diff
@@ -477,11 +478,14 @@ def readchunk(self):
             chunk_data = self.__buffer
         elif self.__position < int(self.length):
             chunk_number = int((received + self.__position) / chunk_size)
-            chunk = self.__chunks.find_one({"files_id": self._id,
-                                            "n": chunk_number},
-                                           session=self._session)
-            if not chunk:
-                raise CorruptGridFile("no chunk #%d" % chunk_number)
+            if self.__chunk_iter is None:
+                self.__chunk_iter = _GridOutChunkIterator(
+                    self, self.__chunks, self._session, chunk_number)
+
+            try:
+                chunk = self.__chunk_iter.next()
+            except StopIteration:
```

Comment:
Is this possible? It looks like _GridOutChunkIterator will never raise StopIteration.

Reply:
`_GridOutChunkIterator` does raise StopIteration after reading the final chunk and checking that no extra chunks exist and I believe I had to make this change to catch a specific test failure when the final chunk in a file was truncated. However, this was on an earlier version of these changes. Now that `_GridOutChunkIterator` checks for missing/empty chunks and raises `CorruptGridFile` itself you are right that StopIteration is impossible here. Also see the coverage report saying this line was never reached: https://s3.amazonaws.com/mciuploads/mongo-python-driver/coverage/a84f50b998ea4614312b8ed13278c52031db2a54/5c7f1e502a60ed235e482752/htmlcov/gridfs_grid_file_py.html#n488